\begin{abstract}
This thesis presents the development of an efficient algorithm that detecting genomic inversions using sequencing data, specifically focusing on long reads and using sample-specific strings to identify the inversions. \\
It begins by introducing the concept of inversions, explaining their biological meaning, and providing a detailed overview of the data structures and definitions that form the basis of the algorithm. \\
The algorithm itself is then thoroughly described, with a complete theoretical analysis of its design, its computational complexity and the logic that lies behind it. \\
Later on, the thesis covers the practical implementation and experimentation, that was performed using the SVDSS tool to validate the algorithm. The results are finally summarized and discussed, demonstrating the algorithm's effectiveness.

\end{abstract}
\addtocounter{page}{0} 